// Copyright 2016 The Fuchsia Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto3";

package cobalt;

// A Cobalt Epoch is a contiguous sequence of days used to aggregate
// observations. Each Report analyzes a set of observations from one epoch.
// Some encodings, such as Forculus, use the current epoch as a parameter.
//
// Each Cobalt |Observation| includes a |day_index| which determines in
// which epoch the Observation falls.
enum EpochType {
  // Each day is a different epoch.
  DAY = 0;
  // A week epoch is a seven day period from  Sunday to Saturday
  WEEK = 1;
  // A month epoch is the set of days in a single month of the Gregorian
  // calendar
  MONTH = 2;
}


///////////////////////////////////////////////////////////////
// Messages that describe the configuration of an encoding.
// These are registered in the Cobalt Config Database and consumed
// by both the Encoder and the Analyzer. Each ObservationPart (see cobalt.proto)
// contains the ID of an EncodingConfig describing how the data
// was encoded.
///////////////////////////////////////////////////////////////

message ForculusConfig {
  uint32 threshold = 1;

  // Forculus threshold encryption is based on the current epoch. For example
  // if epoch_type = WEEK and threshold = 20 then the criteria for being able
  // to decrypt a given ciphertext is that at least 20 different clients
  // all submit observations of that ciphertext tagged with a day_index in
  // the same week, from Sunday to Saturday.
  EpochType epoch_type = 2;
}

message RapporConfig {
  // k = the number of Bloom filter bits. This must satisfy
  // 1 <= h < k < 1024, where h = num_hashes.
  uint32 num_bloom_bits = 1;

  // h = the number of hashes. This must satisfy
  // 1 <= h <  8
  uint32 num_hashes = 2;

  // m = the number of cohorts. This must satisfy
  // 1 <= m < 1024
  uint32 num_cohorts = 3;

  // All probabilities below must be in the range [0.0, 1.0]

  // p = p(a zero bit is changed to a one bit in the IRR)
  float prob_0_becomes_1 = 4;

  // q = p(a one bit remains a one bit in the IRR)
  float prob_1_stays_1 = 5;

  // f = p(a bit is randomly assigned a value in the PRR)
  // Note: Not Implemented in version 0.1 of Cobalt.
  // This value must not be set to a non-zero value.
  float prob_rr = 6;
}

message BasicRapporConfig {
  // All probabilities below must be in the range [0.0, 1.0]

  // p = p(a zero bit is changed to a one bit in the IRR)
  float prob_0_becomes_1 = 1;

  // q = p(a one bit remains a one bit in the IRR)
  float prob_1_stays_1 = 2;

  // f = p(a bit is randomly assigned a value in the PRR)
  // NOTE: PRR is not implemented in Version 0.1 of Cobalt.
  // This value must not be set to a non-zero value.
  float prob_rr = 3;

  // For basic RAPPOR the Encoder needs to know the list of categories.
  // The categories are indexed from 0 to n - 1 where n is the number
  // of categories. The index of a category determines its bit position in
  // the encoding. The categories are listed here in index order.
  // They must be distinct, non-empty strings.
  repeated string category= 7;
}

message EncodingConfig {
  uint32 customer_id = 1;
  uint32 project_id = 2;
  uint32 id = 3;
  oneof config {
    ForculusConfig forculus = 4;
    RapporConfig rappor = 5;
    BasicRapporConfig basic_rappor = 6;
  }
}
